### PR TITLE
Dump of redesigned ONVIF camera DS

### PR DESCRIPTION
**Pull request by jdharms**
Device service has been adapted to model other
existing device services a bit more closely, to
include a pseudo-"plugin" model for integrating
camera- and manufacturer-specific APIs into the DS,
and be compatible with the Fuji release of EdgeX.

### CODE REVIEW

File: internal/pkg/digest/dclient.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	// resp.Body intentionally not closed
+
+	return resp, err
+}
+
+func (dc *DClient) getDigestParts(resp *http.Response) {
+	result := map[string]string{}
+	if len(resp.Header["Www-Authenticate"]) > 0 {
```
Comment (tobiasmo1) [inline internal/pkg/digest/dclient.go]: Case should be `WWW-Authenticate`
ref: https://tools.ietf.org/html/rfc2617

File: internal/pkg/digest/dclient.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+		wantedHeaders := []string{"nonce", "realm", "qop"}
+		responseHeaders := strings.Split(resp.Header["Www-Authenticate"][0], ",")
```
Comment (tobiasmo1) [inline internal/pkg/digest/dclient.go]: Case should be `WWW-Authenticate`

**Comment by jdharms** [comment]
@tobiasmo1 I believe I have addressed the issues you brought up above, including linting, formatting, and a few improvements to the documentation/readme file. 

I'll need to shift my focus for a bit away from this work, but per the ds wg call today, I will take a pass over the new device services document https://wiki.edgexfoundry.org/display/FA/Process+for+new+Device+Services to take a look for any glaring holes.  If you do the same please do let me know of any as you find them.

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,22 @@
+module github.com/edgexfoundry-holding/device-camera-go
+
+require (
+	github.com/BurntSushi/toml v0.3.1
+	github.com/beevik/etree v1.1.0 // indirect
+	github.com/brianolson/cbor_go v1.0.0 // indirect
+	github.com/edgexfoundry/device-sdk-go v0.0.0-20190522161733-b70462a1276d
```
Comment (tobiasmo1) [inline go.mod]: We used this, successfully building (with go 1.13).
+       github.com/edgexfoundry/device-sdk-go v1.0.0
+       github.com/edgexfoundry/go-mod-core-contracts v0.1.0

File: cmd/res/configuration.toml [hunk 4]
```diff
@@ -0,0 +1,66 @@
+[Service]
+Host = "127.0.0.1"
```
Comment (tobiasmo1) [inline cmd/res/configuration.toml]: This is an observation that may be best resolved by documentation or a reference to a general Device Service Troubleshooting section..

When launching EdgeX stack via docker-compose up (edinburgh/compose-files/docker-compose.yml).

Running the service on native host (./run.sh) causes **edgex-core-command** to throw _'Connection Refused'_ errors from the DS.

This is resolved by updating the value in `/cmd/res/configuration.toml` for the service. Change from "127.0.0.1" to docker gateway ip ("172.17.0.1"), so it becomes:

```
[Service]
Host = "172.17.0.1"
...

```

NOTE: As with many persisted artifacts, you may find it critical to first clear the cached mongodb service entries and restart a DS. 

This is easily done by:

a) running the cleanup script (via `developer-scripts/clean_mongo.js`).

b) Modify `developer-scripts/reset-dockers.sh` to include command to enumerate and remove volumes used by EdgeX:
`docker volume rm <edgex volume names>`

c) Or, using the more universal way to cleanup after terminating EdgeX Stack (without requiring mongo client, etc), by dropping the existing docker volumes; e.g., execute 
`docker-compose down -v
`

File: internal/driver/onvifclient.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+	}
+
+	url := uriResponse.MediaUri.Uri
+
+	req, err := http.NewRequest(http.MethodGet, string(url), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.digestClient.Do(req)
```
Comment (tobiasmo1) [inline internal/driver/onvifclient.go]: As discussed, client to be conditional based on support authentication types ('digest', 'basic' or 'none').

File: cmd/res/configuration-driver.toml [hunk 6]
```diff
@@ -0,0 +1,3 @@
+[Camera]
+User = "service"
+Password = "Password!1"
```
Comment (tobiasmo1) [inline cmd/res/configuration-driver.toml]: For initial authentication methods on ONVIF actions. 
```
# Assign AuthMethod to "basic" | "digest" | "none"
AuthMethod = "digest"
```

File: internal/driver/onvifclient.go [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+	dev.Auth(user, password)
+	err := dev.Initialize()
+	if err != nil {
+		lc.Error(fmt.Sprintf("Error initializing ONVIF Client: %v", err.Error()))
+		return nil
+	}
+
+	c.onvifDevice = dev
+
+	c.digestClient = digest.NewDClient(&http.Client{}, user, password)
```
Comment (tobiasmo1) [inline internal/driver/onvifclient.go]: Conditional on requested AuthMethod from conf.

Comment (jdharms) [reply internal/driver/onvifclient.go]: I believe this should always be the digest client--ONVIF itself always uses digest auth.

From the ONVIF spec:

> 5.12.1 Authentication
> The services defined in this standard shall be protected using digest authentication according to [RFC 2617]

**Comment by jdharms** [comment]
Superseded by #11

File: internal/driver/onvifclient.go (hunk 7 above)
Comment (tobiasmo1) [reply internal/driver/onvifclient.go]: Agree. Yes, I put this comment in the wrong place.

**Pull request closed without merging**